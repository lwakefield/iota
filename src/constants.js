export const ELEMENT_NODE = 1
export const TEXT_NODE = 3
export const FORM_ELS = [
  'button',
  'datalist',
  'fieldset',
  'form',
  'input',
  'keygen',
  'label',
  'legend',
  'meter',
  'optgroup',
  'option',
  'output',
  'progress',
  'select',
  'textarea',
]
