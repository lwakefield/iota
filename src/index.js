import {
  observe,
  proxy,
  arrToObj,
  swap,
  max,
} from './util'

const ELEMENT_NODE = 1
const TEXT_NODE = 3
const components = {}

export function app(el, options = {}) {
  const component = new Component(options)
  component.render = Codegen.codegen(el)
  component.mount(el)
  return component
}

export function registerComponent(component) {
  components[component.name] = component
}

export class Component {
  constructor (options = {}) {
    const {data = {}} = options
    this.$data = observe(data, this.update.bind(this))
    this.$el = null
    this._patcher = null
    this.$props = {}

    proxy(this, this.$data)
  }
  mount (el) {
    this.$el = el
    this._patcher = new Patcher(domToVdom(el))
  }
  setProps(props) {
    this.$props = props
  }
  render () {
    throw new Error('Not implemented')
  }
  update () {
    const rendered = this.render.call(
      Object.assign({vnode, tnode}, this.$data, this)
    )
    this._patcher.patch(rendered)
  }
}

export class Codegen {
  static codegen (node) {
    return new Function(
      `with (this) return ${Codegen.codegenNode(node)}`
    )
  }
  static codegenNode (node) {
    if (node.nodeType === TEXT_NODE) {
      return Codegen.codegenTextNode(node)
    } else if (node.nodeType === ELEMENT_NODE) {
      return Codegen.codegenElementNode(node)
    }
    return null
  }
  static codegenTextNode (node) {
    const text = node.textContent
    return `tnode(\`${text}\`)`
  }
  static codegenElementNode (node) {
    const params = [
      `'${node.tagName.toLowerCase()}'`,
      Codegen.codegenAttributes(node),
      Codegen.codegenChildren(node)
    ]
    let code = `vnode(${params.join(',')})`

    const attrs = arrToObj(
      Array.from(node.attributes),
      ({name, value}) => ({[name]: value})
    )
    if (attrs['i-if']) {
      code = `${attrs['i-if']} ? ${code} : null`
    }
    if (attrs['i-for' ]) {
      const [, local, from] = attrs['i-for'].match(/(.*) of (.*)/)
      code = `...${from}.map(${local} => ${code})`
    }
    return code
  }
  static codegenAttributes (node) {
    const props = Array.from(node.attributes)
      .filter(({name}) => name.match(/^:/))
      .map(({name, value}) => `${name.replace(':', '')}: ${value}`)
      .join(',')
    const attrs = Array.from(node.attributes)
      .filter(({name}) =>
        name !== 'i-if' &&
        name !== 'i-for' &&
        !name.match(/^:/)
      )
      .map(({name, value}) => `${name}: \`${value}\``)
    if (props.length) {
      attrs.push(`props: {${props}}`)
    }
    return `{${attrs.join(',')}}`
  }
  static codegenChildren (node) {
    const children = Array.from(node.childNodes).map(Codegen.codegenNode)
    return `[${children.join(',')}]`
  }
}

export class Patcher {
  // Patches nodeB onto nodeA
  constructor (nodeA) {
    this.lastNodeA = nodeA
  }
  patch(nodeA, nodeB) {
    if (arguments.length === 1) {
      this.patch(this.lastNodeA, nodeA)
      this.lastNodeA = nodeA
      return
    }

    // We expect nodeA and nodeB to be of the same type
    if (nodeA.nodeType !== nodeB.nodeType) {
      throw new Error('expected nodes to have the same type')
    }
    nodeB.el = nodeB.el || nodeA.el

    const isComponent = node => (
      node.nodeType === ELEMENT_NODE &&
      !!components[node.tagName]
    )
    const getComponent = node => components[node.tagName]

    if (isComponent(nodeB)) {
      if (!nodeB.component) {
        nodeB.component = new (getComponent(nodeB))
        nodeB.component.mount(nodeA.el)
      }
      nodeB.component.update()
    } else if (nodeA.nodeType === TEXT_NODE) {
      nodeA.el.textContent = nodeB.textContent
    } else if (nodeA.nodeType === ELEMENT_NODE) {
      this.patchAttributes(nodeA, nodeB)
      this.patchChildren(nodeA, nodeB)
    }
  }
  patchAttributes(nodeA, nodeB) {
    // We expect nodeA to have already been mounted
    const attrsA = nodeA.attributes
    const attrsB = nodeB.attributes

    const keys = new Set([...Object.keys(attrsA), ...Object.keys(attrsB)])
    for (const key of keys) {
      if (key in attrsA && !(key in attrsB)) {
        nodeA.el.removeAttribute(key)
      } else if (attrsA[key] !== attrsB[key]) {
        nodeA.el.setAttribute(key, attrsB[key])
      }
    }

    nodeA.attributes = nodeB.attributes
  }
  patchChildren(nodeA, nodeB) {
    const childrenA = nodeA.children
    const childrenB = nodeB.children

    // const getKeyed = node => keyed[`${getTagName(node)}.${getKey(node)}`]
    const nodeTypesMatch = (nodeA, nodeB) => {
      return nodeA.nodeType === nodeB.nodeType &&
        getTagName(nodeA) === getTagName(nodeB)
    }
    // replace a with b
    const replaceNode = (a, b) => a !== b && a.parentNode
      && a.parentNode.replaceChild(b, a)
    // insserts b after a
    const insertAfter = (a, b) => a.parentNode
      && a.parentNode.insertBefore(b, a.nextSibling)
    const len = max(childrenA.length, childrenB.length)
    const getFullKey = node => {
      return node && node.attributes && node.attributes.key ?
        `${getTagName(node)}.${node.attributes.key}` :
        null
    }

    const indexed = new Index()
    for (let i = 0; i < childrenA.length; i++) {
      const child = childrenA[i]
      const key = child.component ?
        child.component.constructor.name :
        getFullKey(child)
      if (key) {
        indexed.queue(key, child)
      }
    }

    function reconcile (childA, childB) {
      const [keyA, keyB] = [getFullKey(childA), getFullKey(childB)]
      const keyedChildA = indexed.dequeue(keyB)
      if (keyedChildA) {
        swap(childrenA, childA, keyedChildA)
        childA && replaceNode(childA.el, keyedChildA.el)
        return keyedChildA
      } else if (keyA && childB) {
        // We might want to use childA at a later point in time...
        childB.el = createElement(childB)
        replaceNode(childA.el, childB.el)
        return childB
      }
      return childA
    }

    for (let i = 0; i < len; i++) {
      let childA = childrenA[i]
      const childB = childrenB[i]

      childA = reconcile(childA, childB)

      if (i > 0 && childA && childA.el && !childA.el.parentNode) {
        // This happens when we remove a keyed node earlier in the loop
        // ie. case 2 in reconcile
        insertAfter(childrenA[i - 1].el, childA.el)
      }

      if (childA && childB) {
        if (nodeTypesMatch(childA, childB)) {
          childB.el = childB.el || childA.el
          this.patch(childA, childB)
        } else {
          childB.el = createElement(childB)
          replaceNode(childA.el, childB.el)
          this.patch(shallowCloneNode(childB), childB)
        }
      } else if (!childA) {
        childB.el = createElement(childB)
        nodeA.el.appendChild(childB.el)
        this.patch(shallowCloneNode(childB), childB)
      } else if (!childB) {
        nodeA.el.removeChild(childA.el)
      }

      childrenA[i] = childB
    }
  }
}

export function vnode (tagName, attributes = {}, children = []) {
  return {tagName, attributes, children, nodeType: ELEMENT_NODE}
}

export function tnode (text) {
  return {textContent: text, nodeType: TEXT_NODE}
}

export function shallowCloneNode (node) {
  const {tagName, attributes, el, textContent, nodeType} = node
  if (node.nodeType === ELEMENT_NODE) {
    return Object.assign({}, {tagName, attributes, el, nodeType, children: []})
  } else if (node.nodeType === TEXT_NODE) {
    return Object.assign({}, {textContent, nodeType, el})
  }
  return null
}

export function getKey (node) {
  if (!node.nodeName) return node.attributes.key || null
  return node._attributes ? node._attributes.key : null
}

export function getTagName (node) {
  return node.tagName
    ? node.tagName.toLowerCase()
    : null
}

export function createElement (node) {
  if (node.nodeType === ELEMENT_NODE) {
    const el = document.createElement(node.tagName)
    const attrs = node.attributes
    for (const key in attrs) {
      el.setAttribute(key, attrs[key])
    }
    return el
  } else if (node.nodeType === TEXT_NODE) {
    return document.createTextNode(node.textContent)
  }
  return null
}

export class Index {
  constructor () {
    this.index = {}
    this.size = 0
  }
  queue (key, val) {
    if (!(key in this.index)) {
      this.index[key] = []
    }
    this.index[key].push(val)
    this.size++
  }
  peek (key) {
    return this.index[key]
  }
  dequeue (key) {
    if (!(key in this.index)) return null
    if (this.index[key].length) this.size--
    return this.index[key].shift() || null
  }
}

export function domToVdom (node) {
  if (node.nodeType === ELEMENT_NODE) {
    const tagName = node.tagName.toLowerCase()
    const attrs = arrToObj(
      Array.from(node.attributes),
      ({name, value}) => ({[name]: value})
    )
    const children = Array.from(node.childNodes).map(domToVdom)
    const n = vnode(tagName, attrs, children)
    n.el = node
    return n
  } else if (node.nodeType === TEXT_NODE) {
    const n = tnode(node.textContent)
    n.el = node
    return n
  }

  return null
}

