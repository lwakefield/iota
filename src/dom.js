import {FORM_ELS, BOOLEAN_ATTRS} from './constants'
// replace a with b
export const replaceNode = (a, b) => a && b && a !== b &&
  a.parentNode && a.parentNode.replaceChild(b, a)
// insserts b after a
export const insertAfter = (a, b) => a.parentNode
  && a.parentNode.insertBefore(b, a.nextSibling)
export const removeNode = (a) => a.parentNode
  && a.parentNode.removeChild(a)

const FORM_EL_REGEX = new RegExp(FORM_ELS.join('|'))
export const isFormEl = node => FORM_EL_REGEX.test(node.tagName.toLowerCase())
export const isBoolAttr = name => BOOLEAN_ATTRS.indexOf(name) !== -1

export function setAttribute (el, name, value) {
  if (isBoolAttr(name) && (value === 'false' || !value)) {
    el.removeAttribute(name)
  } else {
    el.setAttribute(name, value)
  }
}
