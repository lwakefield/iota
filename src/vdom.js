import {ELEMENT_NODE, TEXT_NODE} from './constants'

export function vnode (tagName, options = {}, children = []) {
  return {tagName, options, children, nodeType: ELEMENT_NODE}
}

export function tnode (text) {
  return {textContent: text, nodeType: TEXT_NODE}
}

export function shallowCloneNode (node) {
  const {tagName, options, el, textContent, nodeType} = node
  if (node.nodeType === ELEMENT_NODE) {
    return Object.assign({}, {tagName, options, el, nodeType, children: []})
  } else if (node.nodeType === TEXT_NODE) {
    return Object.assign({}, {textContent, nodeType, el})
  }
  return null
}

export function getTagName (node) {
  return node.tagName
    ? node.tagName.toLowerCase()
    : null
}

export function createElement (node) {
  if (node.nodeType === ELEMENT_NODE) {
    const el = document.createElement(node.tagName)
    const attrs = node.options.attributes || {}
    for (const key in attrs) {
      el.setAttribute(key, attrs[key])
    }
    return el
  } else if (node.nodeType === TEXT_NODE) {
    return document.createTextNode(node.textContent)
  }
  return null
}
