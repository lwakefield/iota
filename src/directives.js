import {isFormEl, isBoolAttr} from './dom'

export const directives = {}

export function registerDirective(directive) {
  directives[directive.name.toLowerCase()] = directive
}
export function unregisterDirective(directiveOrName) {
  if (directiveOrName instanceof String) {
    delete directives[name]
  } else {
    delete directives[(directiveOrName.name || '').toLowerCase()]
  }
}

export default class Directive {
  bind () {}
  update () {}
  unbind () {}
}

export function event(name, value) {
  return {name, value, constructor: Event}
}

export function attr(name, value) {
  return {name, value, constructor: Attribute}
}

export class Attribute extends Directive {
  bind (el, {name, value}) {
    el.setAttribute(name, value)
  }
  update (el, {name, value}, {oldValue}) {
    if (value === oldValue) return

    if (isBoolAttr(name) && (value === 'false' || !value)) {
      el.removeAttribute(name)
    } else {
      el.setAttribute(name, value)
      if (name === 'value' && isFormEl(el)) {
        el.value = value
      }
    }
  }
  unbind (el, {name}) {
    el.removeAttribute(name)
  }
}

export class Event extends Directive {
  constructor () {
    super(...arguments)
    this.handler = []
    this.listener = null
  }
  bind (el, {name, value}) {
    this.handler = value
    this.listener = $event => {
      const result = this.handler($event)
      if (result instanceof Function) {
        result($event)
      }
    }
    el.addEventListener(name, this.listener)
  }
  update (el, {value}) {
    this.handler = value
  }
  unbind(el, {name}) {
    el.removeEventListener(name, this.listener)
  }
}
