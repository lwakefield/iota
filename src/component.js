import Patcher from './patcher'
import {observe, proxy} from './util'
import {vnode, tnode, createElement, shallowCloneNode} from './vdom'
import {replaceNode} from './dom'

export const components = {}

export function registerComponent(component) {
  components[component.name] = component
}
export function unregisterComponent(name) {
  delete components[name]
}

// export function component(options = {}) {
//   const {el, name} = options
//   const component = class extends Component {
//     constructor() {super(options)}
//   }
//   component.name = name
//   component.prototype.render = Codegen.codegen(el)
//   return component
// }

export class Component {
  constructor (options = {}) {
    const {data = {}, methods} = options
    this.$data = observe(data, this.update.bind(this))
    this.$methods = methods
    this.$el = null
    this._patcher = null
    this.$props = {}

    this.vnode = vnode
    this.tnode = tnode

    proxy(this, this.$data)
    const $methods = {}
    for (const name in methods) {
      $methods[name] = methods[name].bind(this)
    }
    proxy(this, $methods)
  }
  mount (el) {
    const rendered = this.render.call(this)
    rendered.el = createElement(rendered)
    this.$el = rendered.el
    replaceNode(el, this.$el)

    this._patcher = new Patcher(shallowCloneNode(rendered))
    this.update()
  }
  setProps(props) {
    this.$props = props
  }
  render () {
    throw new Error('Not implemented')
  }
  update () {
    const rendered = this.render.call(this)
    this._patcher.patch(rendered)
  }
}

