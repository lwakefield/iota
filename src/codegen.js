import {
  TEXT_NODE,
  ELEMENT_NODE,
} from './constants'
import {arrToObj} from './util'
import {isFormEl, isBoolAttr} from './dom'
import {tnode, vnode} from './vdom'
import {attr, event} from './directives'

export function codegen (node) {
  const renderFn = new Function(
    'tnode', 'vnode', 'attr', 'event', 'scope',
    `with (this) return ${codegenNode(node)}`
  )
  return renderFn.bind(this, tnode, vnode, attr, event)
}

export function codegenNode (node) {
  if (node.nodeType === TEXT_NODE) {
    return codegenTextNode(node)
  } else if (node.nodeType === ELEMENT_NODE) {
    return codegenElementNode(node)
  }
  return null
}

export function codegenTextNode (node) {
  const text = node.textContent
  return `tnode(\`${text}\`)`
}

export function codegenElementNode (node) {
  const params = [
    `'${node.tagName.toLowerCase()}'`,
    codegenOptions(node),
    codegenChildren(node)
  ]
  let code = `vnode(${params.join(',')})`

  const attrs = arrToObj(
    Array.from(node.attributes),
    ({name, value}) => ({[name]: value})
  )
  if (attrs['i-if']) {
    code = `${attrs['i-if']} ? ${code} : null`
  }
  if (attrs['i-for' ]) {
    const [, local, from] = attrs['i-for'].match(/(.*) of (.*)/)
    code = `...${from}.map((${local}, $index) => ${code})`
  }
  return code
}

export function codegenOptions (node) {
  const props = []
  const directives = []
  let key = null

  function addDirective (key, val) {
    directives.push(`${key}: ${val}`)
  }

  if (isFormEl(node)) {
    if (/\${.*}/.test(node.getAttribute('value'))) {
      const pointer = node.getAttribute('value').match(/\${(.*)}/)[1]
      addDirective(
        '__formBinding',
        `event('input', $event => ${pointer} = $event.target.value)`
      )
    } else if (/\${.*}/.test(node.getAttribute('checked'))) {
      const pointer = node.getAttribute('checked').match(/\${(.*)}/)[1]
      addDirective(
        '__formBinding',
        `event('change', $event => ${pointer} = $event.target.checked)`
      )
    }
  }

  for (const {name, value} of Array.from(node.attributes)) {
    if (name === 'i-for' || name === 'i-if') continue

    if (name === 'key') {
      key = `key: \`${value}\``
    } else if (name[0] === ':') {
      props.push(`${name.substr(1)}: ${value}`)
    } else if (name[0] === '@') {
      addDirective(
        `'${name}'`,
        `event('${name.substr(1)}', $event => ${value})`
      )
    } else if (isBoolAttr(name) && value === '') {
      addDirective(
        name,
        `attr('${name}', true)`
      )
    } else {
      addDirective(
        name,
        `attr('${name}', \`${value}\`)`
      )
    }
  }

  if (node.attributes['i-for'] && !node.attributes['key']) {
    key = 'key: $index'
  }

  const toAdd = []
  key && toAdd.push(key)
  directives.length && toAdd.push(`directives: {${directives.join(',')}}`)
  props.length && toAdd.push(`props: {${props.join(',')}}`)

  return `{${toAdd.join(',')}}`
}

export function codegenChildren (node) {
  const children = Array.from(node.childNodes).map(codegenNode).filter(v => !!v)
  return `[${children.join(',')}]`
}
