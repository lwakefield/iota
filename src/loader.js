import {expect} from 'chai'
import {jsdom} from 'jsdom'
import {codegenNode} from './codegen'

module.exports = function (content) {
  const doc = jsdom(content)
  const template = doc.querySelector('template')
  const script = doc.querySelector('script')

  // we want to compile the template, then inject it into the script

  const src = codegenNode(template.content)
  return script.innerHTML
}
