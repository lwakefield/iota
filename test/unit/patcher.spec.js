/* eslint-env jest */
import beautify from 'js-beautify'

import Component from '../../src/component'
import Patcher, {Index} from '../../src/patcher'
import {vnode, tnode} from '../../src/vdom'
import {
  htov,
  spy,
  unspy,
  stub,
  unstub,
} from './util'

const assertCode = codeOrNode => expect(
  beautify.html(typeof codeOrNode === 'string' ? codeOrNode : codeOrNode.outerHTML)
).toMatchSnapshot()

describe('Patcher', () => {
  describe('patch', () => {
    it('patches an element', () => {
      const [nodeA, nodeB] = [htov('<div />'), vnode('div')]
      const patcher = new Patcher(nodeA)
      stub(patcher, ['patchChildren', 'patchDirectives'])

      patcher.patch(nodeB)
      assertCode(nodeA.el)
      expect(patcher.patchChildren).toHaveBeenCalledWith(nodeA, nodeB)
      expect(patcher.patchDirectives).toHaveBeenCalledWith(nodeA, nodeB)
    })

    it('patches text', () => {
      const [nodeA, nodeB] = [htov('foo'), tnode('bar')]
      const patcher = new Patcher(nodeA)

      patcher.patch(nodeB)
      expect(nodeA.el.textContent).toEqual(nodeB.textContent)
      expect(patcher.nodeA.el).toBeDefined()
    })

    describe('patches a component', () => {
      class Foo extends Component {}
      beforeEach(() => {
        Component.register('Foo', Foo)
        stub(Foo.prototype, ['mount'])
        spy(Foo.prototype, ['setProps'])
      })
      afterEach(() => {
        Component.unregister('Foo')
        unstub(Foo.prototype, ['mount'])
        unspy(Foo.prototype, ['setProps'])
      })

      it('does a simple patch with props', () => {
        const [nodeA, nodeB] = [
          htov('<div />'),
          vnode('Foo', {props: {foo: 'foobar'}}),
        ]
        const patcher = new Patcher(nodeA)

        const originalEl = nodeA.el
        patcher.patch(nodeB)

        const {component} = nodeA
        expect(component).toBeDefined()
        expect(nodeA.el).toBeDefined()
        expect(component.$el).toBeDefined()
        // Will replace the node on first patch
        expect(nodeA.el).not.toEqual(originalEl)
        expect(component.$el).toEqual(nodeA.el)
        expect(component.$props).toEqual({foo: 'foobar'})
        expect(component.mount).toHaveBeenCalledWith(originalEl)
      })
    })
  })
  describe('patchDirectives', () => {
    const {patchDirectives} = (new Patcher())

    class MockDirective {
      constructor () {
        this.bind = jest.fn()
        this.unbind = jest.fn()
        this.update = jest.fn()
      }
    }

    it('binds', () => {
      const nodeA = htov('<div />')
      const nodeB = vnode('div', {
        directives: {
          foo: {name: ':foo', value: 'bar', constructor: MockDirective},
        }
      })

      patchDirectives(nodeA, nodeB)

      expect(nodeA.options.directives).toEqual(nodeB.options.directives)
      const {instance} = nodeA.options.directives.foo
      expect(instance).toBeDefined()
      expect(instance.bind).toHaveBeenCalledTimes(1)
      expect(instance.bind).toHaveBeenCalledWith(nodeA.el, nodeB.options.directives.foo)
    })

    it('updates', () => {
      const nodeA = htov('<div />')
      nodeA.options.directives[':foo'] = {
        name: ':foo', value: 'bar', constructor: MockDirective, instance: new MockDirective()
      }
      const nodeB = vnode('div', {
        directives: {
          [':foo']: {name: ':foo', value: 'baz', constructor: MockDirective},
        }
      })

      const oldVal = nodeA.options.directives[':foo']
      const {instance} = oldVal

      patchDirectives(nodeA, nodeB)

      expect(nodeA.options.directives).toEqual(nodeB.options.directives)
      expect(nodeA.options.directives[':foo'].instance).toBe(instance)
      expect(instance.update).toHaveBeenCalledTimes(1)
      expect(instance.update).toHaveBeenCalledWith(
        nodeA.el,
        {name: ':foo', value: 'baz'},
        {oldName: ':foo', oldValue: 'bar'}
      )
    })

    it('unbinds', () => {
      const nodeA = htov('<div />')
      nodeA.options.directives[':foo'] = {
        name: ':foo', value: 'bar', constructor: MockDirective, instance: new MockDirective()
      }
      const nodeB = vnode('div', {
        directives: {}
      })

      const oldVal = nodeA.options.directives[':foo']
      const {instance} = oldVal

      patchDirectives(nodeA, nodeB)

      expect(nodeA.options.directives).toEqual(nodeB.options.directives)
      expect(instance.unbind).toHaveBeenCalledTimes(1)
      expect(instance.unbind).toHaveBeenCalledWith(
        nodeA.el,
        {oldName: ':foo', oldValue: 'bar'}
      )
    })
  })

  describe('patchChildren', () => {
    function setup (htmlForNodeA, nodeB) {
      const nodeA = htov(htmlForNodeA)
      const patcher = new Patcher()
      return [nodeA, nodeB, patcher]
    }

    describe('simple functionality', () => {
      beforeEach(() => {
        stub(Patcher.prototype, 'patch')
      })
      afterEach(() => {
        unstub(Patcher.prototype, 'patch')
      })

      it('adds children', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div />',
          vnode('div', {}, [vnode('div')])
        )

        assertCode(nodeA.el)
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
        expect(nodeA.children[0].el).toBeDefined()
      })
      it('removes children', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div><div /></div>',
          vnode('div')
        )

        assertCode(nodeA.el)
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
      })
      it('replace children with non-matching tagNames', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div><div /></div>',
          vnode('div', {}, [vnode('p')])
        )

        assertCode(nodeA.el)
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
        expect(nodeA.children[0].el).toBeDefined()
      })
      it('replace children element node with text node', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div><div /></div>',
          vnode('div', {}, [tnode('hello')])
        )

        assertCode(nodeA.el)
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
        expect(nodeA.children[0].el).toBeDefined()
      })
      it('patches type matching children', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div><div /></div>',
          vnode('div', {}, [vnode('div', {id: 'foo'})])
        )

        assertCode(nodeA.el)
        const firstChildA = nodeA.children[0]
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
        expect(patcher.patch).toHaveBeenCalledWith(firstChildA, nodeB.children[0])
      })
    })

    describe('patching components', () => {
      class Foo extends Component {
        render () {
          return vnode('div', {attributes: {id: 'foo'}})
        }
      }
      beforeEach(() => Component.register('Foo', Foo))
      afterEach(() => Component.unregister('Foo'))

      it('persists components', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div></div>',
          vnode('div', {}, [vnode('Foo'), vnode('Foo')])
        )
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)

        const components = nodeA.children.map(v => v.component)
        expect(components[0]).toBeDefined()
        expect(components[0].$el).toBeDefined()
        expect(components[1]).toBeDefined()
        expect(components[1].$el).toBeDefined()

        const nodeC = vnode('div', {}, [vnode('Foo'), vnode('Foo')])
        patcher.patchChildren(nodeA, nodeC)
        assertCode(nodeA.el)
        expect(nodeA.children[0].component).toBe(components[0])
        expect(nodeA.children[1].component).toBe(components[1])

        const nodeD = vnode('div', {}, [
          vnode('p'), vnode('Foo'), vnode('Foo')
        ])
        patcher.patchChildren(nodeA, nodeD)
        assertCode(nodeA.el)
        expect(nodeA.children[0].tagName).toEqual('p')
        expect(nodeA.children[1].component).toBe(components[0])
        expect(nodeA.children[2].component).toBe(components[1])
      })
      it('persists keyed components', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div></div>',
          vnode('div', {}, [
            vnode('Foo', {key: 0}),
            vnode('Foo', {key: 1}),
            vnode('Foo', {key: 2})
          ])
        )
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)

        const components = nodeA.children.map(v => v.component)
        expect(components.length).toEqual(3)
        components.forEach(component => {
          expect(component).toBeDefined()
          expect(component.$el).toBeDefined()
        })

        const nodeC = vnode('div', {}, [
          vnode('Foo', {key: 1}),
          vnode('Foo', {key: 2}),
          vnode('Foo', {key: 0}),
        ])
        patcher.patchChildren(nodeA, nodeC)
        assertCode(nodeA.el)
        expect(nodeA.children[0].component).toBe(components[1])
        expect(nodeA.children[1].component).toBe(components[2])
        expect(nodeA.children[2].component).toBe(components[0])

        const nodeD = vnode('div', {}, [
          vnode('Foo', {key: 0}),
          vnode('Foo', {key: 2}),
        ])
        patcher.patchChildren(nodeA, nodeD)
        assertCode(nodeA.el)
        expect(nodeA.children.length).toEqual(2)
        expect(nodeA.children[0].component).toBe(components[0])
        expect(nodeA.children[1].component).toBe(components[2])
      })
    })

    describe('patching keyed nodes', () => {
      beforeEach(() => stub(Patcher.prototype, 'patch'))
      afterEach(() => unstub(Patcher.prototype, 'patch'))

      it('patches keyed nodes for the first time', () => {
        const [nodeA, nodeB, patcher] = setup(
          '<div></div>',
          vnode('div', {}, [
            vnode('div', {key: 1}),
            vnode('div', {key: 2}),
            vnode('div', {key: 3}),
          ])
        )
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
      })

      it('patches keyed nodes with no changes', () => {
        const [nodeA, nodeB, patcher] = setup(
          `<div>
            <div key="1"></div>
            <div key="2"></div>
            <div key="3"></div>
          </div>`,
          vnode('div', {}, [
            vnode('div', {key: 1}),
            vnode('div', {key: 2}),
            vnode('div', {key: 3}),
          ])
        )

        const childEls = nodeA.children.map(v => v.el)
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
        nodeA.children.forEach(v => expect(v.el).toBeDefined())
        expect(nodeA.children[0].el).toBe(childEls[0])
        expect(nodeA.children[1].el).toBe(childEls[1])
        expect(nodeA.children[2].el).toBe(childEls[2])
      })

      it('patches keyed nodes with shuffle', () => {
        const [nodeA, nodeB, patcher] = setup(
          `<div>
            <div key="1"></div>
            <div key="2"></div>
            <div key="3"></div>
          </div>`,
          vnode('div', {}, [
            vnode('div', {key: 2}),
            vnode('div', {key: 3}),
            vnode('div', {key: 1}),
          ])
        )
        const childEls = nodeA.children.map(v => v.el)
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
        nodeA.children.forEach(v => expect(v.el).toBeDefined())
        expect(nodeA.children[0].el).toBe(childEls[1])
        expect(nodeA.children[1].el).toBe(childEls[2])
        expect(nodeA.children[2].el).toBe(childEls[0])
      })

      it('patches keyed nodes with non-keyed insertion', () => {
        const [nodeA, nodeB, patcher] = setup(
          `<div>
            <div key="1"></div>
            <div key="2"></div>
            <div key="3"></div>
          </div>`,
          vnode('div', {}, [
            vnode('p'),
            vnode('div', {key: 1}),
            vnode('div', {key: 2}),
            vnode('div', {key: 3}),
          ])
        )
        patcher.patchChildren(nodeA, nodeB)
        assertCode(nodeA.el)
      })
    })
  })
})

describe('Index', () => {
  it('works as intended', () => {
    const index = new Index()
    index.queue('a', 1)
    expect(index.size).toEqual(1)
    expect(index.peek('a')).toEqual(1)
    index.queue('a', 2)
    expect(index.size).toEqual(2)
    expect(index.peek('a')).toEqual(1)

    expect(index.dequeue('a')).toEqual(1)
    expect(index.peek('a')).toEqual(2)
    expect(index.dequeue('a')).toEqual(2)

    expect(index.peek('a')).toEqual(null)
    expect(index.dequeue('a')).toEqual(null)
  })
})
