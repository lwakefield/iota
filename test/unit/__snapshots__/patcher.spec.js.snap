exports[`Patcher patch patches an element 1`] = `"<div></div>"`;

exports[`Patcher patchChildren patching components persists components 1`] = `
"<div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
</div>"
`;

exports[`Patcher patchChildren patching components persists components 2`] = `
"<div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
</div>"
`;

exports[`Patcher patchChildren patching components persists components 3`] = `
"<div>
    <p></p>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
</div>"
`;

exports[`Patcher patchChildren patching components persists keyed components 1`] = `
"<div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
</div>"
`;

exports[`Patcher patchChildren patching components persists keyed components 2`] = `
"<div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
</div>"
`;

exports[`Patcher patchChildren patching components persists keyed components 3`] = `
"<div>
    <div id=\"foo\"></div>
    <div id=\"foo\"></div>
</div>"
`;

exports[`Patcher patchChildren patching keyed nodes patches keyed nodes for the first time 1`] = `
"<div>
    <div></div>
    <div></div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren patching keyed nodes patches keyed nodes with no changes 1`] = `
"<div>
    <div></div>
    <div></div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren patching keyed nodes patches keyed nodes with non-keyed insertion 1`] = `
"<div>
    <p></p>
    <div></div>
    <div></div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren patching keyed nodes patches keyed nodes with shuffle 1`] = `
"<div>
    <div></div>
    <div></div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality adds children 1`] = `"<div></div>"`;

exports[`Patcher patchChildren simple functionality adds children 2`] = `
"<div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality patches type matching children 1`] = `
"<div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality patches type matching children 2`] = `
"<div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality removes children 1`] = `
"<div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality removes children 2`] = `"<div></div>"`;

exports[`Patcher patchChildren simple functionality replace children element node with text node 1`] = `
"<div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality replace children element node with text node 2`] = `"<div>hello</div>"`;

exports[`Patcher patchChildren simple functionality replace children with non-matching tagNames 1`] = `
"<div>
    <div></div>
</div>"
`;

exports[`Patcher patchChildren simple functionality replace children with non-matching tagNames 2`] = `
"<div>
    <p></p>
</div>"
`;
